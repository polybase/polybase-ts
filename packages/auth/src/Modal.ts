export const POLYBASE_AUTH_CSS_PREFIX = 'polybase-auth-'
export const POLYBASE_AUTH_MODAL_ID = 'polybase-auth-modal'
export const POLYBASE_AUTH_MODAL_IFRAME_ID = `${POLYBASE_AUTH_MODAL_ID}-iframe`
export const POLYBASE_DEFAULT_IFRAME_URL = 'https://auth.polybase.xyz'

export class Modal {
  id: string
  iframe: HTMLIFrameElement
  modal: HTMLDivElement

  constructor (id: string, src?: string) {
    this.id = id
    this.iframe = document.createElement('iframe')
    this.iframe.src = src ?? POLYBASE_DEFAULT_IFRAME_URL
    this.modal = document.createElement('div')
    this.modal.onclick = () => {
      this.hide()
    }

    // Append to document
    document.addEventListener('DOMContentLoaded', () => {
      createModal(this.modal, this.iframe, src)
    })
  }

  show = (url?: string) => {
    this.modal.style.display = 'flex'
    if (url && this.iframe.src !== url) {
      this.iframe.src = url
    }
  }

  hide = () => {
    this.modal.style.display = 'none'
  }
}

export function createModal (modal: HTMLDivElement, iframe: HTMLIFrameElement, url?: string) {
  // Create the modal element
  modal.id = 'polybase-auth-modal'
  modal.style.display = 'none'
  modal.classList.add(`${POLYBASE_AUTH_CSS_PREFIX}modal`)

  // Create the modal content container
  const modalContent = document.createElement('div')
  modalContent.classList.add(`${POLYBASE_AUTH_CSS_PREFIX}modal-content`)

  // Create iframe
  iframe.id = POLYBASE_AUTH_MODAL_IFRAME_ID
  iframe.src = url ?? POLYBASE_DEFAULT_IFRAME_URL
  iframe.classList.add(`${POLYBASE_AUTH_CSS_PREFIX}modal-iframe`)

  // Append the iframe to the modal content container
  modalContent.appendChild(iframe)

  // Append the modal content container to the modal element
  modal.appendChild(modalContent)

  // Append the modal to the body
  document.body.appendChild(modal)

  // Append styles
  const css = document.createElement('style')
  css.innerHTML = `
    .${POLYBASE_AUTH_CSS_PREFIX}modal {
      position: fixed; /* positioned relative to the viewport */
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      background-color: rgba(0, 0, 0, 0.75);
      z-index: 1;
      display: flex;
      align-items: center;
      justify-content: center;
      display: none;
    }
    
    .${POLYBASE_AUTH_CSS_PREFIX}modal-content {
      background-color: white;
      border-radius: 10px;
      overflow: hidden;
      height: 400px;
      max-width: 340px;
      width: 100%;
      box-shadow: 0px 3px 6px rgb(0 0 0 / 16%), 0px 3px 6px rgb(0 0 0 / 23%);
    }
    
    .${POLYBASE_AUTH_CSS_PREFIX}modal-iframe{
      width: 100%;
      height: 100%;
      border: none;
    }
  `
  document.body.appendChild(css)

  return modal
}
