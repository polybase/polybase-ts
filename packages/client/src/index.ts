export * from './Polybase'
export * from './Collection'
export * from './Client'
export * from './Query'
export * from './Record'
export * from './Subscription'
export * from './types'
export * from './errors'
