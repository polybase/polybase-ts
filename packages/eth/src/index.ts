export * from './eth'
export * from './sign'
export * from './hash'
