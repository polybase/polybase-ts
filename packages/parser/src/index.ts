/* eslint-disable-next-line import/no-unresolved, import/extensions */
import grammer from './lexer'

export { default as parse } from './parse'
export { default as unparse } from './unparse'
export { filterNodes, mapNodes } from './util'

export type {
  Node,
  BaseNode,
  ArrayNode,
  ObjectNode,
  ObjectKeyValueNode,
  OperatorNode,
  OperatorOpNode,
  FunctionNode,
  NameNode,
  PrimitiveNode,
  NotNode,
  NegateNode,
} from './types'

export default grammer
