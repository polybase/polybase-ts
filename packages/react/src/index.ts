export * from './PolybaseProvider'
export * from './useCollection'
export * from './useCollectionOnce'
export * from './useRecord'
export * from './useRecordOnce'
export * from './usePolybase'
export * from './auth/AuthProvider'
export * from './auth/useAuth'
export * from './auth/useIsAuthenticated'
