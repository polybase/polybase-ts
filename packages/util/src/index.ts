export * from './util'
export * from './deprecated'
export { BytesLike } from '@ethersproject/bytes'
export * as aescbc from './algorithems/aes-cbc'
export * as secp256k1 from './algorithems/secp256k1'
export * as x25519xsalsa20poly1305 from './algorithems/x25519-xsalsa20-poly1305'
export * from './types'
