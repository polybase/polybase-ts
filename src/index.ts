export * from './Spacetime'
export * from './Collection'
export * from './Client'
export * from './types'
