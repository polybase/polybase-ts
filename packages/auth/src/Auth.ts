import { connectToChild, Connection, AsyncMethodReturns } from 'penpal'
import { Modal } from './Modal'

export interface AuthState {
  type: 'metamask' | 'email'
  email?: string | null
  userId?: string | null
  publicKey?: string | null
}

export interface ActionRequestEthPersonalSign {
  type: 'ethPersonalSign',
  data: {
    msg: string
  }
}

export interface ActionRequestSignIn {
  type: 'signIn',
}

export type ActionRequest = ActionRequestEthPersonalSign | ActionRequestSignIn

export interface AuthConfig {
  domain?: string,
  url: string
}

export interface Register {
  domain: string
}

export interface ChildFns {
  register: (register: Register) => Promise<void>
  action: (action: ActionRequest) => Promise<any>
}

export type AuthListener = (state: AuthState | null, auth: Auth) => void

export const defaultConfig = {
  url: 'https://auth.testnet.polybase.xyz',
}

export class Auth {
  config?: AuthConfig
  modal: Modal
  connection: Connection<ChildFns>
  isAuthenticated: boolean
  state: AuthState | null
  loading: boolean
  promise: Promise<AsyncMethodReturns<ChildFns>>
  authUpdateListeners: AuthListener[] = []

  constructor(config?: AuthConfig) {
    this.config = {
      ...defaultConfig,
      ...(config ?? {}),
    }

    this.modal = new Modal(`${Date.now()}`, this.config?.url)

    this.connection = connectToChild({
      // The iframe to which a connection should be made.
      iframe: this.modal.iframe,
      // Methods the parent is exposing to the child.
      methods: {
        onAuthUpdate: (auth: AuthState | null) => {
          this.loading = false
          this.isAuthenticated = !!auth
          if (!isEqual(this.state, auth)) {
            this.authUpdateListeners.forEach((fn) => {
              fn(auth, this)
            })
          }
          this.state = auth
        },
        show: () => {
          this.modal.show()
        },
        hide: () => {
          this.modal.hide()
        },
      },
    })
    this.isAuthenticated = false
    this.state = null
    this.loading = true
    this.promise = this.init()
  }

  init = async () => {
    const child = await this.connection.promise
    await child.register({
      domain: this.config?.domain ?? window.location.origin,
    })
    return child
  }

  signIn = async () => {
    return this.action({
      type: 'signIn',
    })
  }

  ethPersonalSign = async (msg: string) => {
    return this.action({
      type: 'ethPersonalSign',
      data: {
        msg,
      },
    })
  }

  action = async (action: ActionRequest) => {
    return (await this.promise).action(action)
  }

  onAuthUpdate = (listener: AuthListener) => {
    // Add listener
    this.authUpdateListeners.push(listener)
    // Call listener
    listener(this.state, this)
    return () => {
      // Remove listener
      const index = this.authUpdateListeners.indexOf(listener)
      this.authUpdateListeners.splice(index, 1)
    }
  }
}

function isEqual(obj1: any, obj2: any) {
  if (obj1 === obj2) return true
  if (obj1 === null || obj2 === null) return false
  const obj1Keys = Object.keys(obj1)
  const obj2Keys = Object.keys(obj2)
  return obj1Keys.length === obj2Keys.length && obj1Keys.every((key: any) => obj1[key] === obj2[key])
}
