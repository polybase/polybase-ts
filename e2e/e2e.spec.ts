import { Spacetime, CollectionMeta, Collection } from '../src'

jest.setTimeout(7000)

const BASE_API_URL = process.env.E2E_API_URL ?? 'http://localhost:8080'
const API_URL = `${BASE_API_URL}/v0/data`
const wait = (time: number) => new Promise((resolve) => { setTimeout(resolve, time) })

const id = `test-${Date.now()}`

let s: Spacetime

beforeEach(() => {
  s = new Spacetime({
    baseURL: API_URL,
  })
})

test('create collection', async () => {
  const meta: CollectionMeta = {
    id,
    schema: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
        },
      },
    },
    indexes: [{
      fields: [{ field: 'name' }],
    }],
  }

  const res = await s.createCollection(meta)
  expect(res).toBeInstanceOf(Collection)
  expect(res.id).toBe(id)
})

test('add data to collection', async () => {
  const c = await s.collection(id)

  const res = await c.doc('id1').set({
    name: 'Calum',
  })

  expect(res).toEqual({
    data: {
      id: 'id1',
      name: 'Calum',
    },
    block: expect.stringMatching(/^./),
  })
})

test('get data from collection', async () => {
  const c = await s.collection(id)
  const res = await c.doc('id1').get()

  expect(res).toEqual({
    data: {
      id: 'id1',
      name: 'Calum',
    },
    // block: expect.stringMatching(/^./),
  })
})

test('list data from collection', async () => {
  const c = await s.collection(id)
  await c.doc('id2').set({
    name: 'Sally',
  })

  const res = await c.get()

  expect(res).toEqual([{
    data: {
      id: 'id1',
      name: 'Calum',
    },
    // block: expect.stringMatching(/^./),
  }, {
    data: {
      id: 'id2',
      name: 'Sally',
    },
  }])
})

test('list data with where clause', async () => {
  const c = await s.collection(id)
  await c.doc('id3').set({
    name: 'Sally',
  })

  const res = await c.where('name', '==', 'Sally').get()

  expect(res).toEqual([{
    data: {
      id: 'id2',
      name: 'Sally',
    },
    // block: expect.stringMatching(/^./),
  }, {
    data: {
      id: 'id3',
      name: 'Sally',
    },
  }])
})

test('list data with snapshot', async () => {
  const c = await s.collection(id)
  const spy = jest.fn()
  const q = c.where('name', '==', 'Calum')

  const unsub = q.onSnapshot(spy)
  await wait(2000)

  expect(spy).toBeCalledTimes(1)
  expect(spy).toBeCalledWith([{
    data: {
      id: 'id1',
      name: 'Calum',
    },
  }])

  // await c.doc('id4').set({
  //   name: 'Calum',
  // })

  // await wait(3000)

  // expect(spy).toBeCalledTimes(2)
  // expect(spy).toBeCalledWith([{
  //   data: {
  //     id: 'id1',
  //     name: 'Calum',
  //   },
  // }, {
  //   data: {
  //     id: 'id4',
  //     name: 'Calum',
  //   },
  // }])

  unsub()
})
