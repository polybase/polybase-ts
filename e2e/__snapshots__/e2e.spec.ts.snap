// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signing 1`] = `"Request failed with status code 401"`;
